Three more biogo examples

aboffin opened the pull request:
Hi,

This is related to issue: https://github.com/biogo/biogo/issues/27

I have included three additional examples that demonstrate some uses of biogo FASTA file manipulation.

This includes filtering contigs based on length (FastaLenFilter), a simple converter that reads FASTA multiple sequence alignment and converts it to sequential PHYLIP format (FastaToPhylip) and finally a simple "shredder" that splits genomic contigs into fragments of specific window length (SplitGenome). I have included test input and output files for all three.

I hope to spend more time adding more examples, but I wanted to submit these for comments and suggestions.

Thank you in advance for your help,
Senthil

aboffin committed:
moved from biogoTut

aboffin committed:
use command-line or test files

aboffin committed:
use command-line or test files

aboffin committed:
smaller test files

aboffin committed:
avoid temp variables

aboffin committed:
fixed issue with split-join small fragments

kortschak commented on FastaLenFilter/FastaLenFilter.go:
```diff
@@ -0,0 +1,63 @@
+import (
+	"fmt"
+	"io"
+	"github.com/biogo/biogo/alphabet"
```
Please use goimports import formatting:

```
	"fmt"
	"io"
	"os"

	"github.com/biogo/biogo/alphabet"
	"github.com/biogo/biogo/io/seqio/fasta"
	"github.com/biogo/biogo/seq/linear"
```

kortschak commented on FastaLenFilter/FastaLenFilter.go:
```diff
@@ -0,0 +1,63 @@
+var (
+	inf     = flag.String("inf", "test.fna", "input filename")
+	outf    = flag.String("outf", "test_gt2500bp.fna", "output filename")
+	minLen = flag.Int("minLen", 2500, "minimum sequence length cut-off (bp)")
```
Please `gofmt` or `goimports`.

kortschak commented on FastaLenFilter/FastaLenFilter.go:
```diff
@@ -0,0 +1,63 @@
+	}
+	f, err := os.Open(*inf)
+	  if err != nil {
+		  fmt.Fprintf(os.Stderr, "open input FASTA file: %v.", err)
```
`log.Fatalf("failed to open input fasta file: %v", err)` and delete following line.

kortschak commented on FastaLenFilter/FastaLenFilter.go:
```diff
@@ -0,0 +1,63 @@
+	}
+	w := fasta.NewWriter(of, 60)
+	defer of.Close()
+	for {
```
This can be more idiomatically said using the `seqio.Scanner` type.

```
sc := seqio.NewScanner(in)
for sc.Read() {
	s := sc.Seq()
	if s.Len() > *min {
		_, err := w.Write(s)
		if err != nil {
			log.Fatalf("failed to write sequence %q: %v", s.ID(), err)
		}
	}
}
err := sc.Error()
if err != nil {
	log.Fatalf("failed during read: %v", err)
}
```

Note that `_, err := w.Write(s)` can be spelled `_, err := fmt.Fprintf(out, "%60a\n", s)`. When writing to stderr, this is easier, though potentially slower.

kortschak commented on FastaLenFilter/FastaLenFilter.go:
```diff
@@ -0,0 +1,63 @@
+		  os.Exit(1)
+	  }
+	defer f.Close()
+	dnaf := fasta.NewReader(f, linear.NewSeq("", nil, alphabet.DNA))
```
s/dnaf/in/g

kortschak commented on FastaLenFilter/FastaLenFilter.go:
```diff
@@ -0,0 +1,63 @@
+	  }
+	defer f.Close()
+	dnaf := fasta.NewReader(f, linear.NewSeq("", nil, alphabet.DNA))
+	of, err := os.Create(*outf)
```
s/of/out/g

kortschak commented on FastaToPhylip/FastaToPhylip.go:
```diff
@@ -0,0 +1,93 @@
+
+import (
+	"fmt"
+	"github.com/biogo/biogo/alphabet"
```
Imports.

kortschak commented on SplitGenome/SplitGenome.go:
```diff
@@ -0,0 +1,95 @@
+						ff := fs{
+							fe{s: i, e: i + winLen},
+						}
+						if err := sequtils.Stitch(d, s, ff); err == nil {
```
```
err := sequtils.Stitch(d, s, ff)
if err != nil {
	continue
}
...
```

kortschak commented on SplitGenome/SplitGenome.go:
```diff
@@ -0,0 +1,95 @@
+						if err := sequtils.Stitch(d, s, ff); err == nil {
+							d.Desc = fmt.Sprintf("%v_%v", s.Desc, i)
+							if _, err = w.Write(d); err != nil {
+								fmt.Fprintf(os.Stderr, "write FASTA record :%v", err)
```
This message does not say that something went wrong. Prefix with "failed to".

kortschak commented on SplitGenome/SplitGenome.go:
```diff
@@ -33,8 +33,8 @@ type fs []feat.Feature
 var (
-	inf     = flag.String("inf", "bd.fasta", "input filename")
-	outf    = flag.String("outf", "split_bd.fasta", "output filename")
+	inf     = flag.String("inf", "test.fna", "input filename")
```
Have empty default and test for zero value - fail if zero.

kortschak commented on SplitGenome/SplitGenome.go:
```diff
@@ -2,7 +2,11 @@
+// Example: if a contig is 14kb long, split into 5 and 9 kb instead of 5, 5 
+// and 4 kb fragments, see:
+// https://www.ncbi.nlm.nih.gov/pubmed/19698104
+
```
Delete blank line.

kortschak commented on SplitGenome/SplitGenome.go:
```diff
@@ -2,7 +2,11 @@
+// Split genomic contigs into fragments of a given size window (winLen).
+// Example: if a contig is 14kb long, split into 5 and 9 kb instead of 5, 5 
+// and 4 kb fragments, see:
+// https://www.ncbi.nlm.nih.gov/pubmed/19698104
```
This link is not wildly informative.

kortschak commented on SplitGenome/SplitGenome.go:
```diff
@@ -70,25 +75,45 @@ func main() {
-					}
-				} else {
-					for i := 0; i < len(s.Seq); i = i + *winLen {
+			switch {
```
The logic here needs much more explanation if this is going to be an example.

aboffin committed:
use seqio.Scanner

aboffin committed:
imports fixed, use seqio.Scanner

aboffin committed:
use seqio.NewScanner

aboffin committed:
added comments, use seqio.NewScanner, clearer variable naming

aboffin committed:
use seqio.NewScanner, clearer variable names, comments, error messages

aboffin committed:
simplified logic, use seqio.NewScanner

aboffin committed:
gofmt, use seqio.Scanner, use stdin stdout defaults

aboffin commented:
@kortschak Thanks for your input. Based on your suggestions, I have made the following changes:

1. Use gofmt to format import and code
2. Use consistent naming (r for reader, w for writer, sc for scanner, etc.)
3. Use stdin and stdout where possible in place of hard-coded input and output file names
4. SplitGenome: clarified logic, added comments.

I will submit a new pull request.

kortschak commented:
Please note that there is no need to close PRs that are subject to change requests. The PR is against the HEAD of the PR branch. So we just work until we get something that can be merged.

aboffin commented:
@kortschak I am sorry for closing this pull request. I have reopened it (and closed the newer pull request that discusses the same issues). Sorry again for the confusion!

aboffin committed:
added FastaStats

kortschak reviewed (changes requested):
As a general comment, please use snake_case for file names. This is the Go convention.

I'm sorry I have not been able to get to the review yet, I am very busy right now with other things.

aboffin committed:
removed unnecessary types in FastaStats

aboffin commented:
@kortschak I will work on the suggested changes. I understand you are busy, I will check again with you after a few days. Thank you!

aboffin committed:
removed unnecessary types in FastaStats

aboffin committed:
removed temp variables

aboffin committed:
removed temp variables

aboffin committed:
removed temp variables

aboffin committed:
minor edits

aboffin committed:
clearer comparison to check N50

aboffin committed:
clearer compa                         rison to check N50

aboffin committed:
renamed files to camelCase

aboffin committed:
shortened file names and in lower case

aboffin committed:
shortened file names and in lower case

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Convert a multiple-sequence alignment in FASTA to
```
```
// fastatophy converts a multiple-sequence alignment in FASTA to
//  PHYLIP (sequential) format.
```

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+)
+
+func main() {
+	var (
```
Please move the flag vars out to package scope as a single block. Drop out `totSeq` and `alnLen`.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+		flag.Usage()
+		os.Exit(0)
+	}
+	// open input (first pass) to get totSeq and alnLen
```
Please write comments as complete sentences with leading capital and final full stop (also insert a blank line before this comment since it's a separate stanza).

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+		os.Exit(1)
+	}
+	defer out.Close()
+	totSeq = 0
```
s/totSeq/n/g

`var n, alnLen int` and insert blank line before.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+		alnLen = s.Len()
+		totSeq = totSeq + 1
+	}
+	alnCounts := fmt.Sprintf("%d %d\n", totSeq, alnLen)
```
`fmt.Fprintf(out, "%d %d\n", n, alnLen)` and delete following line.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	alnCounts := fmt.Sprintf("%d %d\n", totSeq, alnLen)
+	io.WriteString(out, alnCounts)
+	// open input (second pass) to read sequences and write to output
```
Proper sentence and preceding blank line.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+	in, err := os.Open(*inf)
+	if err != nil {
+		log.Fatalf("open FASTA file (1st pass): %v.\n", err)
+		os.Exit(1)
```
Delete - this is dead code.

kortschak replied:
Ahh, I see - the `os.Exit(1)` is dead, not the entire conditional. `log.Fatal*` calls `os.Exit(1)`.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+	out, err := os.Create(*outf)
+	if err != nil {
+		log.Fatalf("open PHYLIP file: %v.\n", err)
+		os.Exit(1)
```
Same here.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+	in, err = os.Open(*inf)
+	if err != nil {
+		log.Fatalf("open FASTA file (2nd pass): %v.", err)
+		os.Exit(1)
```
Dead code.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+	alnCounts := fmt.Sprintf("%d %d\n", totSeq, alnLen)
+	io.WriteString(out, alnCounts)
+	// open input (second pass) to read sequences and write to output
+	in, err = os.Open(*inf)
```
Instead of reopening, `_, err = os.Seek(0, os.SEEK_SET)`.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+		os.Exit(1)
+	}
+	defer in.Close()
+	r = fasta.NewReader(in, linear.NewSeq("", nil, alphabet.Protein))
```
Insert blank line before this.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+		log.Fatalf("open FASTA file (2nd pass): %v.", err)
+		os.Exit(1)
+	}
+	defer in.Close()
```
No longer necessary after use of seek.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+				s.Name(), alnLen)
+		}
+		if len(s.Name()) > 10 {
+			log.Printf("WARNING: Sequence ID %s is longer than 10 characters.\n",
```
Don't break this line.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+		alnRow := fmt.Sprintf("%s %v\n", s.Name(), s.Seq)
+		io.WriteString(out, alnRow)
+		if s.Len() != alnLen {
+			log.Printf("WARNING: Length of sequence %s is different than %d.\n",
```
Don't break this line. Also for my English's sake, s/different than/different to/,

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+	for sc.Next() {
+		s := sc.Seq().(*linear.Seq)
+		alnRow := fmt.Sprintf("%s %v\n", s.Name(), s.Seq)
+		io.WriteString(out, alnRow)
```
`fmt.Fprinf(out, "%s %v\n", s.Name(), s.Seq)` and delete WriteString call.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+	for sc.Next() {
+		s := sc.Seq()
+		alnLen = s.Len()
+		totSeq = totSeq + 1
```
`n++`

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,82 @@
+	sc := seqio.NewScanner(r)
+	for sc.Next() {
+		s := sc.Seq()
+		alnLen = s.Len()
```
This assignment is unused - presumably there should be a check here.

kortschak replied:
*ping*

kortschak replied:
*ping*

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+		r       *fasta.Reader
+		w       *fasta.Writer
+		err     error
+		inf     = flag.String("inf", "", "input contig file name to be fragmented. Defaults to stdin.")
```
Flags above please.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+		fmt.Fprintln(os.Stderr, "Reading sequences from stdin.")
+		r = fasta.NewReader(os.Stdin, t)
+	} else if in, err = os.Open(*inf); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v.", err)
```
Use `log.Fatalf("failed to open %q: %v", *inf, err)` and drop `os.Exit` call.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Filter sequences that are above a length cut-off
```
Final full stop.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+
+	t := linear.NewSeq("", nil, alphabet.DNA)
+	if *inf == "" {
+		fmt.Fprintln(os.Stderr, "Reading sequences from stdin.")
```
Omit - unix Rule of Silence.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+		os.Exit(1)
+	} else {
+		defer in.Close()
+		fmt.Fprintf(os.Stderr, "Reading sequence from `%s'.\n", *inf)
```
Omit.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	if *outf == "" {
+		fmt.Fprintln(os.Stderr, "Writing output to stdout.")
```
Omit.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+		fmt.Fprintf(os.Stderr, "Writing output to `%s'.\n", *outf)
+	}
+	defer out.Close()
+	w = fasta.NewWriter(out, 60)
```
Blank line before.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+		fmt.Fprintln(os.Stderr, "Writing output to stdout.")
+		out = os.Stdout
+	} else if out, err = os.Create(*outf); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v.", err)
```
This should be fatal - report as above with modification.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,76 @@
+	} else if out, err = os.Create(*outf); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v.", err)
+	} else {
+		fmt.Fprintf(os.Stderr, "Writing output to `%s'.\n", *outf)
```
Omit else block.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+// Example: given a 14kb long contig,  generate fragments of size 5kb and
+// 9kb (rather than 5kb, 5kb and 4kb).
+// Add sequence position information to old headers to get new ones.
+
```
Delete blank line.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Split sequence that is above the min cut-off length to generate fragments
```
seqsplit splits sequences that are above...

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+
+// Split sequence that is above the min cut-off length to generate fragments
+// such that each fragment falls in the range: window < fragment < 2*window
+// (i.e. 5 to 10 kb).  To achieve this, split each contig into window-sized
```
To achieve this each contig is split into window-sized... and the remainder is added to the last window...

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+// such that each fragment falls in the range: window < fragment < 2*window
+// (i.e. 5 to 10 kb).  To achieve this, split each contig into window-sized
+// fragments and add the remainder to the last window-sized fragment.
+// Example: given a 14kb long contig,  generate fragments of size 5kb and
```
Add

```
//
```

before this line and remove double space after the comma.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+// fragments and add the remainder to the last window-sized fragment.
+// Example: given a 14kb long contig,  generate fragments of size 5kb and
+// 9kb (rather than 5kb, 5kb and 4kb).
+// Add sequence position information to old headers to get new ones.
```
Clarify this sentence please.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+		r       *fasta.Reader
+		w       *fasta.Writer
+		err     error
+		inf     = flag.String("inf", "", "input contig file name to be fragmented. Defaults to stdin.")
```
Flags in var block outside main.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+
+	t := linear.NewSeq("", nil, alphabet.DNA)
+	if *inf == "" {
+		fmt.Fprintln(os.Stderr, "Reading sequences from stdin.")
```
Omit.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+		fmt.Fprintln(os.Stderr, "Reading sequences from stdin.")
+		r = fasta.NewReader(os.Stdin, t)
+	} else if in, err = os.Open(*inf); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v.", err)
```
`log.Fatalf`

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+		os.Exit(1)
+	} else {
+		defer in.Close()
+		fmt.Fprintf(os.Stderr, "Reading sequence from `%s'.\n", *inf)
```
Omit.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	if *outf == "" {
+		fmt.Fprintln(os.Stderr, "Writing output to stdout.")
```
Omit.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+		fmt.Fprintln(os.Stderr, "Writing output to stdout.")
+		out = os.Stdout
+	} else if out, err = os.Create(*outf); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v.", err)
```
`log.Fatalf`

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+	} else if out, err = os.Create(*outf); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v.", err)
+	} else {
+		fmt.Fprintf(os.Stderr, "Writing output to `%s'.\n", *outf)
```
Omit block.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+		fmt.Fprintf(os.Stderr, "Writing output to `%s'.\n", *outf)
+	}
+	defer out.Close()
+	w = fasta.NewWriter(out, 60)
```
Swap order of these two lines so the `w` is declared int the scan block stanza.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+		startPos, endPos := 0, 0
+		switch {
+		case len(next.Seq) < *min:
+			// discard contigs below cut-off size limit
```
Proper sentence please.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+			// discard contigs below cut-off size limit
+			continue
+		case len(next.Seq) >= 2*(*window):
+			// example: from 27582bp contig, get all 5000 bp sized fragments till
```
Proper sentences please.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+				startPos = i * (*window)
+				endPos = startPos + (*window)
+				ff := fs{
+					fe{s: startPos, e: endPos},
```
`ff := fs{{s: startPos, e: endPos}}`

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+					continue
+				}
+				// add seq locations to header
+				curr.Desc = fmt.Sprintf("%v_%v-%v", next.Desc, startPos,
```
Don't break this line.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+				}
+			}
+			// example: write last 7582 fragment (5000 + 2582)
+			ff := fs{
```
`ff := fs{{s: endPos, e: endPos + (*window) + remainder}}`

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+				fmt.Fprintf(os.Stderr, "failed to write remainder fragment: %v", err)
+			}
+		default:
+			// contig is of desired size range
```
Capital and full stop.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+			if err != nil {
+				continue
+			}
+			curr.Desc = fmt.Sprintf("%v_%v-%v", next.Desc, endPos,
```
Don't break this line.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+			}
+			err := sequtils.Stitch(curr, next, ff)
+			if err != nil {
+				continue
```
I would have thought this was a panic.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,139 @@
+			if _, err = w.Write(next); err != nil {
+				fmt.Fprintf(os.Stderr, "failed to write contig: %v", err)
+			}
+
```
Delete blank lines here.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+// file (default stdin), useful for analyzing metrics of microbial genome
+// assemblies or metagenome "bins". It prints: the total no. of sequences,
+// Assembly size (total length of all sequences), Min, Max, Avg and N50
+
```
Delete blank line.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Calculate and print sequence statistics from a multi-FASTA DNA sequence
```
seqstats calculates and prints sequence statistics...

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+// license that can be found in the LICENSE file.
+
+// Calculate and print sequence statistics from a multi-FASTA DNA sequence
+// file (default stdin), useful for analyzing metrics of microbial genome
```
s/, useful for/. It is useful for/

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+		err     error
+		b       binStats
+		seqlens []int
+		inf     = flag.String("inf", "", "input contig file, defaults to stdin")
```
Flags outside main please.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+
+	t := linear.NewSeq("", nil, alphabet.DNA)
+	if *inf == "" {
+		fmt.Fprintln(os.Stderr, "Reading sequences from stdin.")
```
Omit.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+		fmt.Fprintln(os.Stderr, "Reading sequences from stdin.")
+		r = fasta.NewReader(os.Stdin, t)
+	} else if in, err = os.Open(*inf); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v.", err)
```
`log.Fatalf`

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+		os.Exit(1)
+	} else {
+		defer in.Close()
+		fmt.Fprintf(os.Stderr, "Reading sequence from `%s'.\n", *inf)
```
Omit.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+
+	sc := seqio.NewScanner(r)
+	b.Name = strings.Split(path.Base(*inf), ".")[0]
+	b.Min = 1000000000 // similar: int(^uint(0) >> 1)
```
Use max int rather than an arbitrary large value.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+	if err != nil {
+		log.Fatalf("failed during read: %v", err)
+	}
+	// sort in descending order of lengths
```
Insert blank line. Capital and full stop.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,101 @@
+	}
+	// sort in descending order of lengths
+	sort.Sort(sort.Reverse(sort.IntSlice(seqlens)))
+	// csum stores the cumulative sequence lengths
```
Capital and full stop.

kortschak reviewed (changes requested):
Please gofmt your code (probably with `-S`).

I think these examples could do with more commentary throughout.

aboffin committed:
incorporated suggestions, commentaries pending

aboffin committed:
moved variable declarations

aboffin committed:
added commentaries; formatted with gofmt -s; incorporated all suggested changes

aboffin committed:
added commentaries; formatted with gofmt -s; incorporated all suggested changes

aboffin commented:
@kortschak 
Thank you for reviewing! I have made the requested changes, formatted the code using "gofmt -s" and also added commentaries where they seemed appropriate. Let me know if you have any further comments.

aboffin committed:
added commentaries; formatted with gofmt -s; incorporated all suggested changes

aboffin committed:
minor: remove newlines near end of file

aboffin committed:
fixed capitalization in comments

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,74 @@
+
+// fastatophy converts a multiple-sequence alignment in FASTA to
+// PHYLIP (sequential) format.
+
```
Delete blank line - this breaks godoc.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,74 @@
+		os.Exit(0)
+	}
+
+	in, _ := os.Open(*inf)
```
This assumes `*inf != ""`. Test that and `log.Fatalf(...` if not - also document in flag message.

aboffin replied:
Unlike the other three examples, 'fastatophy' does not
accept files from stdin or write output to stdout. I have
added error checking to inf and outf that was present
in an earlier version. I had taken them out because you
mentioned it as 'dead code', but they probably are needed?

kortschak replied:
Check that `*inf != ""` before trying to open it. Then print `flag.Usage` if it does. This is nicer for the user than just dumping an error to a log file.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,74 @@
+	defer in.Close()
+	r := fasta.NewReader(in, linear.NewSeq("", nil, alphabet.Protein))
+
+	out, _ := os.Create(*outf)
```
Same here.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,74 @@
+	}
+	fmt.Fprintf(out, "%d %d\n", n, alnLen)
+
+	// Read input FASTA file from start to write sequences to output
```
Full stop.

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,78 @@
+	"github.com/biogo/biogo/seq/linear"
+)
+
+// Flags assign input, output file names and minimum length cut-off to their default
```
This is documentation for the Go standard library. I don't think we need that, but rather comments on bíogo-specific things.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,74 @@
+		s := sc.Seq().(*linear.Seq)
+		fmt.Fprintf(out, "%s %v\n", s.Name(), s.Seq)
+		if s.Len() != alnLen {
+			log.Printf("WARNING: Length of sequence %s is different to %d.\n", s.Name(), alnLen)
```
Should this be fatal?

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,74 @@
+	fmt.Fprintf(out, "%d %d\n", n, alnLen)
+
+	// Read input FASTA file from start to write sequences to output
+	in.Seek(0, io.SeekStart)
```
This can error. Check the error.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,74 @@
+		if len(s.Name()) > 10 {
+			log.Printf("WARNING: Sequence ID %s is longer than 10 characters.\n", s.Name())
+		}
+
```
Delete blank line.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,74 @@
+		}
+
+	}
+
```
Here too.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,133 @@
+// added to the last window-sized fragment. Add sequence position
+// information to the contig FASTA header to generate unique new headers
+// for the split sequences.
+
```
This blank line breaks the godoc, it must be a commented blank line.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,133 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// seqsplit splits sequence that is above the min cut-off length to
```
...splits sequences that are above the...

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,133 @@
+
+// seqsplit splits sequence that is above the min cut-off length to
+// generate fragments such that each fragment falls in the size range:
+// window < fragment < 2*window (i.e. 5 to 10 kb).  To achieve this each
```
Single space after punctuation please.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,133 @@
+			// Discard contigs below the cut-off size limit.
+			continue
+		case len(next.Seq) >= 2*(*window):
+			// For example, split a 27582 bp contig into 5000 bp sized fragments till
```
This is unclear. If you are explaining something explain it and possibly add an example if it helps, do not explain by merely providing an example.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,133 @@
+				if err != nil {
+					panic(err)
+				}
+				// add seq locations to header
```
Proper sentence.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,133 @@
+					fmt.Fprintf(os.Stderr, "failed to write window-sized fragment: %v", err)
+				}
+			}
+			// Continuing on the above example, write the remaining 7582 (5000 +
```
Please don't break expressions like this.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+var (
+	inf  = flag.String("inf", "", "input contig file, defaults to stdin")
```
Please change the flag string "inf" to "in". The name "inf" is already heavily loaded.

aboffin replied:
I have changed "inf" to "ctgf".

kortschak replied:
"in" please.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	sc := seqio.NewScanner(r)
+	// Get the basename of the file and remove the extension.
```
This comment is not helpful.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,109 @@
+	// Get the basename of the file and remove the extension.
+	// Example: "/path/to/infile.fasta" -> "infile.fasta" -> "infile"
+	b.Name = strings.Split(path.Base(*inf), ".")[0]
+	// Assign b.Min to the largest possible integer value. 
```
Same here - also note that this code is not safe for 32 bit architectures. Use a safe definition (one is provided in the util package or preferably you can just copy the definition here) and use an appropriately named const rather than a comment.

kortschak reviewed (commented):
Please take a look at https://www.lysator.liu.se/c/pikestyle.html, particularly the section on comments.

aboffin committed:
pikestyle comments, error-checking, safe MaxInt

aboffin committed:
remove lagging blank lines

aboffin committed:
remove lagging blank lines

aboffin commented:
@kortschak

1. Thank you for suggesting the `pikestyle' guide. I
have read it many times, but now I applied some of its
suggestions. I have extensively rewritten the comments,
hopefully they are clear and succinct now.

2. I have fixed the blank lines at the end of files using
`gofmt -s'.

3. I made sure that comments are full sentences and end
with a period and that they do not break godoc formatting.

4. I have used a constant MaxInt, derived from uint instead
of hard coding it.

5. I have added a padding function in `fastatophy'
that appends spaces to identifiers to make them exactly
10 characters. Long Identifiers are truncated to 10
characters. These changes are to adhere to the "strict"
PHYLIP format.

6. I have commented on two issues specifically, let me know
if I understood your comments correctly and the changes
look alright.

Thank you,

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -20,55 +19,89 @@ import (
 
+// padding returns requested number of spaces as a
+// concatenated string.
+func padding(n int) (spaces string) {
```
A much simpler way of writing this is 

```
const padding = "          "
```

and move it down to where it is used.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -20,55 +19,89 @@ import (
+			strictName = s.Name()[:10]
+			log.Printf("Identifier: %s was truncated to 10 characters\n", s.Name())
+		} else {
+			strictName = s.Name() + padding(10-len(s.Name()))
```
```
const padding = "          " // Ten spaces.
strictName = s.Name() + padding[:10-len(s.Name())]
```

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,143 @@
+// seqsplit splits contig sequences that are above a
+// minimum cut-off length to generate fragments such that
+// each fragment falls in the size range:
+// window < fragment < (2*window).
```
Inset these equations by one more space to make them a `<pre>`.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,143 @@
+// each fragment falls in the size range:
+// window < fragment < (2*window).
+// This is achieved as follows, calculate:
+// remainder = length % window and
```
These two as well.

kortschak reviewed (commented):
I really would like in files to be given the CLI flag name "in".

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -20,55 +19,89 @@ import (
-
+	// Reinitialize to read from the start of the FASTA file
+	// and write the alignment section to the PHYLIP file.
+	if _, err := in.Seek(0, io.SeekStart); err != nil {
```
```
_, err = in.Seek(0, io.SeekStart)
if err != nil {
    log.Fatalf("seek failed: %v", err)
}
```

aboffin committed:
in, out flags; const padding; inline equations

aboffin commented:
@kortschak 
1. I have changed flag names for input and output filenames to "in" and "out", respectively, in all the 4 examples.
2.  `fastatophy.go`:
     * I have added a space to make the equations inline.
     * Thank you for the neat replacement for the padding function, I have used `const` instead, as you suggested.
     * Regarding the `alnLen = s.Len()` not being used and needing a check there, doesn't the block that's shown below, take care of it? I added it based on your suggestion of using `sc.Error()` after `b.Max = s.Len()` in `seqstats.go`.
```
	err = sc.Error()
	if err != nil {
		log.Fatalf("failed during read: %v", err)
	}
```
Thank you and have a good weekend.

aboffin committed:
minor comment capitalization

aboffin committed:
≤ symbol instead of <

aboffin committed:
Merge commit 'f9db85054228224c36fbf6b5c6af5cb5acedf89c' of https://github.com/aboffin/examples
add greater than sign

kortschak commented:
>Regarding the `alnLen = s.Len()` not being used and needing a check there, doesn't the block that's shown below, take care of it?

Sort of. A usage text for an absent filename is usual. Erroring with `failed to open FASTA file: open : no such file or directory` is much less informative that a list of expected parameters and a usage message (note that as it stands, the filename being absent only shows up as an odd whitespace).

Please make the check prior to the `os.Open` calls (preferrably just after the help flag check - which flag package does for you for).

aboffin commented:
@kortschak Thank you for your comment. `fastatophy` unlike other three examples, expects `test.aln` as the default input file (and not stdin). So if `test.aln` does not exist, it shows the following error (rather than a simple " " space character with error message):
```
2017/06/08 20:58:43 failed to open FASTA file: open test.aln: no such file or directory
fastatophy: exit 1
```
Also, `fastatophy -help` outputs:
```
Usage of fastatophy:
  -help
    	help prints this message
  -in string
    	input FASTA filename (default "test.aln")
  -out string
    	output PHYLIP filename (default "test.phy")
```
If I understand correctly, if `test.aln` is present, it is read and if there are any errors in the FASTA records, they are caught by `sc.Error()`. Any hinters or pointers on what is missing is really appreciated, thank you in advance.

kortschak commented:
I'd prefer a `""` default and test for the zero value than a "test.aln"/"test.phy" default (or maybe even allowing the default to use stdin and stdout). Can you explain why you have decided to go for different behaviour with this program compared to the other two?

aboffin commented:
@kortschak I used a specific input file because the input file `test.aln` is read twice, once for getting the alignment length and total number of sequences and once for copying the actual alignments (FASTA sequences) to the output file. Since the file is reset by `io.SeekStart` to read the file again from start the second time, stdin cannot be used here. I used a specific output file to stay consistent with using a specific input file (rather than mixing a combination of a specific input file and using stdout as default).

kortschak commented:
Fair enough. Please do use a "" default though.

aboffin committed:
fixed default file names in fastatophy; added sc.Error() check to seqsplit

aboffin commented:
@kortschak Thank you, I have fixed the file names as suggested.

aboffin committed:
minor edit: comment Capitalization

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,99 @@
+		os.Exit(0)
+	}
+
+	in, err := os.Open(*inf)
```
Please check that `*inf != ""`.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,99 @@
+	defer in.Close()
+	r := fasta.NewReader(in, linear.NewSeq("", nil, alphabet.Protein))
+
+	out, err := os.Create(*outf)
```
Same here.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,99 @@
+)
+
+func main() {
+	var (
```
Declare these just before they are used.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,99 @@
+	// Read all FASTA records to get total number of sequences
+	// (n) and length of each sequence. alnLen stores the
+	// sequence length of the last record.
+	n = 0
```
`var n, alnLen int`

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,99 @@
+	sc := seqio.NewScanner(r)
+	for sc.Next() {
+		s := sc.Seq()
+		alnLen = s.Len()
```
This still needs a check. Just assigning to a var in a loop is a bad code smell - I would check that they agree.

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,99 @@
+		s := sc.Seq().(*linear.Seq)
+		// Assert that each sequence in the multiple-sequence
+		// alignment is of equal length.
+		if s.Len() != alnLen {
```
If you do the check above this extra work is not needed (unix rule of repair).

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,73 @@
+		defer in.Close()
+		r = fasta.NewReader(in, t)
+	}
+
```
`var out *os.File`

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,73 @@
+		os.Exit(0)
+	}
+
+	t := linear.NewSeq("", nil, alphabet.DNA)
```
```
var in *os.File
var err error
```

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,73 @@
+	}
+	defer out.Close()
+
+	w = fasta.NewWriter(out, 60)
```
`w := fasta.NewWriter(out, 60)`

kortschak commented on seqlen/seqlen.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+func main() {
+	var (
```
Declare just before use.

kortschak commented on seqsplit/seqsplit.go:
```diff
@@ -0,0 +1,147 @@
+)
+
+func main() {
+	var (
```
Same here.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+func main() {
+	var (
```
And here.

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,108 @@
+	Size    int
+	Min     int
+	Max     int
+	Avg     int
```
Avg should be a float64

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,108 @@
+		}
+		csum = seqlens[i] + csum
+	}
+	b.Avg = b.Size / b.totSeqs
```
`b.Avg = float64(b.Size) / float64(b.totSeqs)`

kortschak commented on seqstats/seqstats.go:
```diff
@@ -0,0 +1,108 @@
+// any extension and other reported statistics in bp
+// (base pairs).
+type binStats struct {
+	Name    string // From input filename (empty, if stdin).
```
What's the rationale behind the field exportation rules here?

aboffin committed:
field names; declare var before use; check var in for loopxterm

aboffin commented:
@kortschak 
1. Moved variable declarations before their first use.
2. Added check to compare sequence lengths in fastatophy for loop.
3. Changed binStats field names in seqstats (all in small-case).
4. Changed avg to float64 in seqstats.
Thanks!

aboffin committed:
field names; declare var before use; check var in for loop

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,112 @@
+	var r *fasta.Reader
+	var err error
+	t := linear.NewSeq("", nil, alphabet.Protein)
+	if *inf == "" {
```
This should happen before here, just after the help check, and exit(1) if `*inf == ""`.

```
var (
	inf  = flag.String("in", "", "input FASTA filename (required)")
	outf = flag.String("out", "", "output PHYLIP filename")
	help = flag.Bool("help", false, "help prints this message")
)

func main() {
	flag.Parse()
	if *help {
		flag.Usage()
		os.Exit(0)
	}
	if *inf == "" {
		flag.Usage()
		os.Exit(1)
	}

	in, err := os.Open(*inf)
	if err != nil {
		log.Fatalf("failed to open FASTA file %q: %v", *inf, err)
	}
	defer in.Close()
	r := fasta.NewReader(in, 	linear.NewSeq("", nil, alphabet.Protein))
	...
```

aboffin committed:
moved checking

aboffin committed:
moved empty files check

kortschak commented:
Please make the changes I wrote in https://github.com/biogo/examples/pull/3#discussion_r122564173 (including the equivalent for the outfile). There is no need for the var decl, and t should be declared closer to use (certainly not prior to checks as it is now) and the quoted file names should be emitted for errors.

aboffin committed:
fixed per comment 3

aboffin committed:
fixed per comment 3

aboffin commented:
I am sorry I was in a hurry and didn't see your input closely. Now, I have made the changes that you mentioned in comment 3.

aboffin committed:
move t, r close to use

kortschak commented on fastatophy/fastatophy.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+var (
+	inf  = flag.String("in", "", "input FASTA filename")
```
Last change. Please add " (required)" to each of in and out.

aboffin committed:
added required

kortschak commented:
Please send a PR adding yourself to A+C at https://github.com/biogo/biogo

aboffin commented:
@kortschak Thank you, I just opened a PR (biogo/biogo). Can I close this PR (biogo/examples) now?

kortschak commented:
It's merged, no closing is necessary.
